### ISSUE TITLE
Assuming that username == doc._id constrains how you can use the sl_users couch database.

### ISSUE THREAD

**Issue by tohagan** [colinskow/superlogin#56]
Normally a single couch database contains a mixture of document types but sadly this appears to be rather difficult to do with the current `superlogin` structure. I'm basing this on a quick look at auth views like this one ...  

      username: function(doc) {
         emit(doc._id, null);
      },

Please correct me Colin if I missed something here but if I'm right can we can perhaps find some workarounds for now and maybe include some refactoring to support mixed doc types in the user database in the future roadmap?  Of course you may have had some other reasons I'm not aware of so please share them.  I've seen one fork of this project that I think got the complete wrong message about CouchDb and went off to create a database per record type. I think they just  blindly followed your pattern.

Apologies for my next Couch 101 rant ... I know it's all rather basic stuff but I'm unsure if these reasons were missed.

_Why & How to mix doc types in one database ..._

Maintaining a mixture of records in the one db instance also has many benefits including: Simpler/safer db syncing, using single updates via _bulk_docs and single gets via _all_docs, being able to watch all change with a single change feed and leverage many map/reduce view patterns that need to emit a mixture of record types (and often efficiently reduce and cache/index the result).  In my current app I want to do this with the user docs but can't since they are in a different sluser database where _id values might collide unless prefixed (but this would break the current views that assume `username == doc._id`).

A common convention to distinguish document types is to place a prefix on _id values and/or to add a `type` attribute.  The _id prefix means you can then use `startkey` and `endkey` in either a view or _all_docs queries to perform efficient indexed lookups for the subset of record of the given type.

**Comment by tohagan** [comment]
Hmmm ... Maybe I just need to chat with you sometime.  I think theses username are the session generated ones so probably can ignore my rant.

**Comment by colinskow** [comment]
CouchDB databases are very cheap. I never contemplated a compelling use case where there would be an advantage to mixing other types of docs into the `sl_users` database. Any custom object that has a 1:1 relationship with a user, you can safely put inside the user doc.

If you have a use case where there is a huge advantage to mix other documents into `sl_users` tell me about it, and I'll see what we can do to accommodate.

**Comment by colinskow** [comment]
If you want to chat, I am in #SuperLogin on the [PouchDB Slack](http://slack.pouchdb.com). Afternoon California time works best. Let me know when and we can set a time.

**Comment by tohagan** [comment]
Yes I understand disk space cost is very small (16k for an empty db).  I understand the security model reasons for having per user dbs.  I've spent the last 10 months studying everything I can find about CouchDb and Pouch. 

In my case I need to create view(s) that emit/reduce user and group objects (many to many) in the one view so these object types must share the same database. My users also have different roles per group and may eventually have additional attributes on this M:M relation. I will likely use "per user" databases as well but groups and many other shared data objects of my schema won't fit here. I will need to support replication of these and many other shared object types to other couch servers (not pouch) and having multiple replications that don't preserve object insert sequence order would be painful. I also want to be able to perform single updates with multiple doc types using bulk_update (like a transaction) which I can't do if the db is split into multiple dbs. I may also need to watch changes to these shared objects in a single change feed. I also want to only create a single validation doc in this db to manage the schema (not split it up into "per object type" databases).

So in short there are many reasons to use a single couch db for shared data. I'm sure other devs will hit the same issues. This generally means you need to use _id prefixes to avoid key collisions and ensure efficient per type queries using startkey/endkey based of the prefix which  avoids overheads of maintaining a view index for each object type.  

Cheers,
Tony

**Comment by colinskow** [comment]
To summarize our Slack discussion, I am going to check for `type === 'user'` in the CouchDB views, and we will add an optional prefix to the user `_id`. We will also provide a conversion script to prefix the users in the database.

**Comment by tohagan** [comment]
Thanks Colin.
